### PR TITLE
[BEAM-7746] Get mypy passing on runners.worker

### PR DESCRIPTION
**Pull request by chadrik**
Ok, I'm back on the horse, but who knows for how long.  

I reviewed the last PR that stalled out and the recommendation there was to focus on a single module or package.   Let me know if the scope is too large on this one.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x ] Update `CHANGES.md` with noteworthy changes.
 - [ x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by chadrik** [comment]
R: @robertwb 

There are 8 errors remaining which I could use a Beam expert to help me resolve.

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 1]
```diff
@@ -920,7 +928,7 @@ def reset(self):
   def process_bundle(self, instruction_id):
     # type: (str) -> Tuple[List[beam_fn_api_pb2.DelayedBundleApplication], bool]
 
-    expected_inputs = []
+    expected_input_ops = []
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:931]: This needed to be renamed because the variable was reused below for a different type

File: sdks/python/apache_beam/runners/worker/data_plane.py [hunk 2]
```diff
@@ -303,9 +334,10 @@ def inverse(self):
 
   def input_elements(self,
       instruction_id,  # type: str
-      unused_expected_inputs=None,   # type: Collection[str]
+      unused_expected_inputs,   # type: Any
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/data_plane.py:337]: This method was incompatible with its super type.  If it is `None` it invalidates the Liskov substitution principle.  

It's more accurate to say `Any` here, since the method doesn't care, and this keeps the method compatible with its super type.   

I checked all uses of `input_elements` and I did not see any case where it was called with only one arg.

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 3]
```diff
@@ -1344,7 +1356,8 @@ def create_deprecated_read(
     consumers  # type: Dict[str, List[operations.Operation]]
 ):
   # type: (...) -> operations.ReadOperation
-  source = iobase.SourceBase.from_runner_api(parameter.source, factory.context)
+  source = iobase.BoundedSource.from_runner_api(
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:1359]: This shouldn't change the behavior of the code, it merely indicates to the type system that we expect `BoundedSource`, since `SourceBase` is not sufficient here (we need `default_output_coder()`)

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 4]
```diff
@@ -1125,15 +1134,15 @@ def shutdown(self):
 class ExecutionContext(object):
   def __init__(self):
     self.delayed_applications = [
-    ]  # type: List[Tuple[operations.DoOperation, common.SplitResultType]]
+    ]  # type: List[Tuple[operations.DoOperation, common.SplitResultResidual]]
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:1137]: this was just wrong. there is no `common.SplitResultType`

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 5]
```diff
@@ -705,10 +713,11 @@ def __init__(self,
     self._key_coder = key_coder
     self._window_coder = window_coder
     # A mapping of {timer_family_id: TimerInfo}
-    self._timers_info = {}
-    self._all_states = {}  # type: Dict[tuple, userstate.RuntimeState]
+    self._timers_info = {}  # type: Dict[str, TimerInfo]
+    self._all_states = {}  # type: Dict[tuple, FnApiUserRuntimeStateTypes]
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:717]: `userstate.RuntimeState` does not have the necessary `commit()` method, and there is no common base class for these 4 types that has all of the necessary methods.  When dealing with a reasonably limited number of possibilities with unclear parentage, a `Union` is a good solution.

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 6]
```diff
@@ -725,7 +726,7 @@ def __init__(self, key_coder_impl, window_coder_impl):
     self._tag_coder_impl = StrUtf8Coder().get_impl()
 
   def encode_to_stream(self, value, out, nested):
-    # type: (dict, create_OutputStream, bool) -> None
+    # type: (userstate.Timer, create_OutputStream, bool) -> None
```
Comment (chadrik) [inline sdks/python/apache_beam/coders/coder_impl.py:729]: as far as I can tell, this was just wrong

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 7]
```diff
@@ -1070,23 +1078,24 @@ def delayed_bundle_application(self,
...
                         ):
     # type: (...) -> beam_fn_api_pb2.BundleApplication
-    return self.construct_bundle_application(op, None, primary.primary_value)
+    assert op.input_info is not None
+    return self.construct_bundle_application(
+        op.input_info, None, primary.primary_value)
 
   def construct_bundle_application(self,
-                                   op,  # type: operations.DoOperation
+                                   op_input_info,  # type: operations.OpInputInfo
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:1093]: Instead of passing around a `DoOperation` I switched this to a `OpInputInfo` because the latter is all that we need, and doing so let's us avoid having to assert that it is not `None` everywhere that it's used.

File: sdks/python/apache_beam/utils/urns.py [hunk 8]
```diff
@@ -177,7 +178,7 @@ def to_runner_api(self, context):
 
   @classmethod
   def from_runner_api(cls, fn_proto, context):
-    # type: (beam_runner_api_pb2.FunctionSpec, PipelineContext) -> Any
+    # type: (Type[RunnerApiFnT], beam_runner_api_pb2.FunctionSpec, PipelineContext) -> RunnerApiFnT
```
Comment (chadrik) [inline sdks/python/apache_beam/utils/urns.py:181]: This change indicates that we should return an instance of the class that this is called from, though it provides no runtime guarantees of this.

File: sdks/python/tox.ini [hunk 9]
```diff
@@ -27,6 +27,8 @@ select = E3
 
 # Shared environment options.
 [testenv]
+# allow apps that support color to use it.
+passenv=TERM
```
Comment (chadrik) [inline sdks/python/tox.ini:31]: Now you can see mypy color output from tox!  Let me know if you want me to do this in another PR.

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 10]
```diff
@@ -947,11 +955,11 @@ def process_bundle(self, instruction_id):
       # (transform_id, timer_family_id).
       data_channels = collections.defaultdict(
           list
-      )  # type: DefaultDict[data_plane.GrpcClientDataChannel, List[str]]
+      )  # type: DefaultDict[data_plane.GrpcClientDataChannel, List[Union[str, Tuple[str, str]]]]
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:958]: This can in fact contain items which are `Tuple[str, str]`.  see:

```python
        data_channels[self.timer_data_channel].extend(
            list(self.timers_info.keys()))
```

File: sdks/python/apache_beam/runners/worker/data_plane.py [hunk 11]
```diff
@@ -218,7 +249,7 @@ class DataChannel(with_metaclass(abc.ABCMeta, object)):  # type: ignore[misc]
   @abc.abstractmethod
   def input_elements(self,
                      instruction_id,  # type: str
-                     expected_inputs,  # type: Collection[str]
+                     expected_inputs,  # type: Sized
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/data_plane.py:252]: we just call `len(expected_inputs)`, so we just need to know this implements `__len__`

File: sdks/python/apache_beam/runners/worker/data_plane.py [hunk 12]
```diff
@@ -243,7 +274,7 @@ def output_stream(
       instruction_id,  # type: str
       transform_id  # type: str
   ):
-    # type: (...) -> ClosableOutputStream
+    # type: (...) -> SizeBasedBufferingClosableOutputStream
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/data_plane.py:277]: `ClosableOutputStream` seems to not actually be used anywhere any more, because `ClosableOutputStream.create()` returns `SizeBasedBufferingClosableOutputStream`.  `ClosableOutputStream` does not implement the necessary `maybe_flush` method that's required everywhere we were referring to `ClosableOutputStream`

File: sdks/python/apache_beam/utils/sentinel.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+import enum
+
+
+class Sentinel(enum.Enum):
+  """
+  A type-safe sentinel class
+  """
+
+  sentinel = object()
```
Comment (chadrik) [inline sdks/python/apache_beam/utils/sentinel.py:30]: see this [stackoverflow post](https://stackoverflow.com/questions/57959664/handling-conditional-logic-sentinel-value-with-mypy) for an explanation of why I chose this design for a type-safe sentinel.

File: sdks/python/apache_beam/runners/worker/statecache.py [hunk 14]
```diff
@@ -44,15 +48,17 @@ class Metrics(object):
...
 
   def initialize(self):
+    # type: () -> None
+
     """Needs to be called once per thread to initialize the local metrics cache.
     """
     if hasattr(self._context, 'metrics'):
       return  # Already initialized
-    self._context.metrics = collections.defaultdict(
-        int)  # type: DefaultDict[Hashable, int]
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/statecache.py:55]: Removed this annotation because it's invalid:   you can't add annotations to an attribute of another class (non-self attribute).  I've noticed there's definitely a need for a typed version of `threading.local`.

File: sdks/python/apache_beam/runners/worker/log_handler.py [hunk 15]
```diff
@@ -156,9 +174,11 @@ def _write_log_entries(self):
         done = True
         log_entries.pop()
       if log_entries:
-        yield beam_fn_api_pb2.LogEntry.List(log_entries=log_entries)
+        yield beam_fn_api_pb2.LogEntry.List(
+            log_entries=cast(List[beam_fn_api_pb2.LogEntry], log_entries))
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/log_handler.py:178]: we have to cast here because `log_entries` started out life as a `List[Union[..., Sentinel]]`, and just above this line we checked for the sentinel and popped it out, but mypy can't track that.

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 16]
```diff
@@ -1619,7 +1633,7 @@ def _create_pardo_operation(
       consumers,
       output_tags)
   if pardo_proto and pardo_proto.restriction_coder_id:
-    result.input_info = (
+    result.input_info = operations.OpInputInfo(
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:1636]: Making this a NamedTuple added a bit of sanity, and should be completely safe.

File: sdks/python/apache_beam/runners/worker/sdk_worker.py [hunk 17]
```diff
@@ -174,11 +191,14 @@ def __init__(self,
...
-                process_bundle_descriptor_id=id))
-    )  # type: Mapping[str, beam_fn_api_pb2.ProcessBundleDescriptor]
+
+    def default_factory(id):
+      # type: (str) -> beam_fn_api_pb2.ProcessBundleDescriptor
+      return self._control_stub.GetProcessBundleDescriptor(
+          beam_fn_api_pb2.GetProcessBundleDescriptorRequest(
+              process_bundle_descriptor_id=id))
+
+    self._fns = KeyedDefaultDict(default_factory)
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/sdk_worker.py:201]: changing this to a function lets us add annotations which silences a mypy error.  mypy can now detect the key/value types of `KeyedDefaultDict`

File: sdks/python/apache_beam/runners/worker/sdk_worker.py [hunk 18]
```diff
@@ -464,16 +496,17 @@ class SdkWorker(object):
...
+               log_lull_timeout_ns=None,  # type: Optional[int]
               ):
+    # type: (...) -> None
     self.bundle_processor_cache = bundle_processor_cache
     self.state_cache_metrics_fn = state_cache_metrics_fn
     self.profiler_factory = profiler_factory
     self.log_lull_timeout_ns = (
         log_lull_timeout_ns or DEFAULT_LOG_LULL_TIMEOUT_NS)
-    self._last_full_thread_dump_secs = 0
+    self._last_full_thread_dump_secs = 0.0
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/sdk_worker.py:509]: this attribute eventually gets set to the value of `time.time()` which is a `float`.

File: sdks/python/apache_beam/coders/coder_impl.py (hunk 6 above)
Comment (robertwb) [reply sdks/python/apache_beam/coders/coder_impl.py:729]: I think it used to be correct back when timers were being implemented. This code changed a couple of months ago too.

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 19]
```diff
@@ -119,10 +125,10 @@ class RunnerIOOperation(operations.Operation):
 
   def __init__(self,
                name_context,  # type: Union[str, common.NameContext]
-               step_name,
+               step_name,  # type: Any
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:128]: Is this not str or Optional[str]?

File: sdks/python/apache_beam/runners/worker/data_plane.py (hunk 12 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/worker/data_plane.py:277]: We're also thinking about adding a time-based one. 

Let's add a no-op maybe_flush method to the baseclass and keep ClosableOutputStream everywhere.

File: sdks/python/apache_beam/runners/worker/bundle_processor.py (hunk 10 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/worker/bundle_processor.py:958]: Yep, this changed a couple of months ago. It'll be good to finally have these type annotations checked.

File: sdks/python/apache_beam/runners/worker/bundle_processor.py (hunk 7 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/worker/bundle_processor.py:1093]: Sounds good to me.

File: sdks/python/apache_beam/runners/worker/data_plane.py [hunk 20]
```diff
@@ -81,7 +85,11 @@
 
 class ClosableOutputStream(OutputStream):
   """A Outputstream for use with CoderImpls that has a close() method."""
-  def __init__(self, close_callback=None):
+  def __init__(
+      self,
+      close_callback=None  # type: Optional[Optional[Callable[[bytes], None]]]
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/data_plane.py:90]: Why the double Optional (here and elsewhere below)?

File: sdks/python/apache_beam/runners/worker/data_plane.py (hunk 11 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/worker/data_plane.py:252]: Don't we call `__contains__` as well? I'd rather keep it more fully typed.

File: sdks/python/apache_beam/runners/worker/data_plane.py [hunk 21]
```diff
@@ -331,6 +369,7 @@ def add_to_inverse_output(timer):
                 is_last=False))
 
     def close_stream(timer):
+      # type: (bytes) -> None
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/data_plane.py:372]: I wonder if we should call this encoded_timer[s]?

**Review by robertwb — APPROVED** [review]
Looks good to me. 

(I have to say that having to type methods like `__hash__` and `__repr__` are unfortunate, and in my book the `-> None`s are a bit noisy, but if that's what it takes to make mypy happy it's worth it.

File: sdks/python/apache_beam/runners/worker/data_plane.py (hunk 20 above)
Comment (chadrik) [reply sdks/python/apache_beam/runners/worker/data_plane.py:90]: mistake.  fixed.

File: sdks/python/apache_beam/runners/worker/bundle_processor.py (hunk 19 above)
Comment (chadrik) [reply sdks/python/apache_beam/runners/worker/bundle_processor.py:128]: It looks like this argument isn't used at all any more.

The inheritance tree looks like this:

- `operations.Operation`: does not accept a `step_name` argument
  - `RunnerIOOperation`: calls `super`. does nothing with `step_name`
    - `DataOutputOperation`: does not override `__init__`
    - `DataInputOperation`:  calls `super`. does nothing with `step_name`

File: sdks/python/apache_beam/runners/worker/data_plane.py (hunk 12 above)
Comment (chadrik) [reply sdks/python/apache_beam/runners/worker/data_plane.py:277]: fixed.

File: sdks/python/apache_beam/runners/worker/data_plane.py (hunk 11 above)
Comment (chadrik) [reply sdks/python/apache_beam/runners/worker/data_plane.py:252]: we only call `__len__`, but I'm happy to leave it as a collection.  updated.

**Comment by robertwb** [comment]
> 23:36:24 apache_beam/runners/worker/data_plane.py:714: error: Incompatible return value type (got "Optional[GrpcClientDataChannel]", expected "GrpcClientDataChannel")  [return-value]

Sure this shouldn't return optional as well? 

> 23:36:24 apache_beam/runners/worker/operations.py:722: error: Argument 1 to "append" of "list" has incompatible type "Tuple[DoOperation, Iterable[SplitResultResidual]]"; expected "Tuple[DoOperation, SplitResultResidual]"  [arg-type]

self.dofn_runner.process does return an `Iterable[SplitResultResidual]`

> 23:36:24 apache_beam/runners/worker/bundle_processor.py:241: error: Signature of "try_split" incompatible with supertype "Operation"  [override]

The supertype returns Optional[Any], which should be compatible with this. 

> 23:36:24 apache_beam/runners/worker/bundle_processor.py:978: error: "Operation" has no attribute "add_timer_info"  [attr-defined]
> 23:36:24 apache_beam/runners/worker/bundle_processor.py:990: error: "Operation" has no attribute "process_timer"  [attr-defined]

These will always be DoOperations.

>23:36:24 apache_beam/runners/worker/bundle_processor.py:1568: error: Argument 5 to "StateBackedSideInputMap" has incompatible type "Coder"; expected "WindowedValueCoder"  [arg-type]

I think this will always hold, we just don't know it due to having deserialized it with the generic mechanisms.

**Comment by robertwb** [comment]
> 23:36:24 apache_beam/runners/worker/sdk_worker.py:665: error: Argument "process_bundle_progress" to "InstructionResponse" has incompatible type "ProcessBundleProgressMetadataResponse"; expected "Optional[ProcessBundleProgressResponse]"  [arg-type]
> 23:36:24 apache_beam/runners/worker/sdk_worker.py:667: error: Argument "monitoring_info" to "ProcessBundleProgressMetadataResponse" has incompatible type "List[MonitoringInfo]"; expected "Optional[Mapping[str, Any]]"  [arg-type]

@lukecwik this looks wrong, wonder why it's not failing at runtime. Maybe short-ids aren't actually used yet?

File: sdks/python/apache_beam/runners/worker/sdk_worker.py [hunk 22]
```diff
@@ -622,10 +659,12 @@ def process_bundle_progress_metadata_request(self,
...
                                               ):
+    # type: (...) -> beam_fn_api_pb2.InstructionResponse
     return beam_fn_api_pb2.InstructionResponse(
         instruction_id=instruction_id,
-        process_bundle_progress=beam_fn_api_pb2.
+        process_bundle_progress_metadata=beam_fn_api_pb2.
         ProcessBundleProgressMetadataResponse(
+            # FIXME: incompatible type "List[MonitoringInfo]"; expected "Optional[Mapping[str, Any]]"
             monitoring_info=SHORT_ID_CACHE.getInfos(
                 request.monitoring_info_id)))
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/sdk_worker.py:669]: Note this change and the FIXME.  I'm not sure what to do about this. 

I got complaints that `InstructionResponse(process_bundle_progress=...)` was receiving the wrong type.  I'm not sure whether this is the right resolution.

Also, `ProcessBundleProgressMetadataResponse(monitoring_info=...)`is also receiving the wrong type.  It's supposed to be a mapping not a list.  Should I change `SHORT_ID_CACHE.getInfos` to return a dictionary? 

Do you know if this code is being hit?  If so, any theories on how it's working?

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 23]
```diff
@@ -966,6 +976,7 @@ def process_bundle(self, instruction_id):
         output_stream = self.timer_data_channel.output_timer_stream(
             instruction_id, transform_id, timer_family_id)
         timer_info.output_stream = output_stream
+        # FIXME: how do we know that self.ops[transform_id] is a DoOperation?
         self.ops[transform_id].add_timer_info(timer_family_id, timer_info)
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:980]: `BundleProcessor.ops` is `OrderedDict[str, operations.Operation]`.  Should it be `OrderedDict[str, operations.DoOperation]`?  Or do the transform ids in `BundleProcessor.timers_info` all point to `DoOperations`?

**Comment by chadrik** [comment]
> apache_beam/runners/worker/operations.py:722

**Comment by chadrik** [comment]
> > 23:36:24 apache_beam/runners/worker/data_plane.py:714: error: Incompatible return value type (got "Optional[GrpcClientDataChannel]", expected "GrpcClientDataChannel")  [return-value]
> 
> Sure this shouldn't return optional as well?

When I do that, it just causes errors elsewhere.  I couldn't find a solution.  I think there's some implicit guarantee going on here.  I'd like to identify where it is and add an assert, a cast, and/or a comment, but I don't know the system well enough.

> > 23:36:24 apache_beam/runners/worker/operations.py:722: error: Argument 1 to "append" of "list" has incompatible type "Tuple[DoOperation, Iterable[SplitResultResidual]]"; expected "Tuple[DoOperation, SplitResultResidual]"  [arg-type]
> 
> self.dofn_runner.process does return an `Iterable[SplitResultResidual]`

yeah, the problem is that `self.execution_context.delayed_applications` is `List[Tuple[operations.DoOperation, common.SplitResultResidual]]`, and we're trying to append to it.   Should this be `Union[Iterable[SplitResultResidual], SplitResultResidual]`, or is something wrong here?

> > 23:36:24 apache_beam/runners/worker/bundle_processor.py:241: error: Signature of "try_split" incompatible with supertype "Operation"  [override]
> 
> The supertype returns Optional[Any], which should be compatible with this.

The number of arguments is different. 

> 
> > 23:36:24 apache_beam/runners/worker/bundle_processor.py:978: error: "Operation" has no attribute "add_timer_info"  [attr-defined]
> > 23:36:24 apache_beam/runners/worker/bundle_processor.py:990: error: "Operation" has no attribute "process_timer"  [attr-defined]
> 
> These will always be DoOperations.

Does `create_execution_tree` return `OrderedDict[str, operations.DoOperation]`? 


> 
> > 23:36:24 apache_beam/runners/worker/bundle_processor.py:1568: error: Argument 5 to "StateBackedSideInputMap" has incompatible type "Coder"; expected "WindowedValueCoder"  [arg-type]
> 
> I think this will always hold, we just don't know it due to having deserialized it with the generic mechanisms.

I changed `BeamTransformFactory.get_input_coders` to return `Dict[str, coders.WindowedValueCoder]`.  Does that seem right?

**Comment by chadrik** [comment]
We're getting merge conflicts.  I'm thinking of just ignoring the remaining errors, so that we can just get this merged.  I can add a FIXME referencing BEAM-7746.   Thoughts?

The one I'm most concerned about is `ProcessBundleProgressMetadataResponse`, which seems like an actual bug.  @lukecwik any thoughts on that one?

File: sdks/python/apache_beam/runners/worker/sdk_worker.py (hunk 22 above)
Comment (lukecwik) [reply sdks/python/apache_beam/runners/worker/sdk_worker.py:669]: This code is not being used currently and it should be a mapping.

Yes, please update getInfos to return that mapping.

**Comment by robertwb** [comment]
Yes, let's fix `ProcessBundleProgressMetadataResponse`, drop some TODOs, enable what checks we can, and get this merged.

**Comment by chadrik** [comment]
Done.

**Comment by chadrik** [comment]
I think the only failures remaining are just flaky tests.

**Comment by robertwb** [comment]
Yeah, PTransformTest.test_flatten_no_pcollections is flaky. Lint, however, isn't.

**Comment by chadrik** [comment]
whoops.  ok, all lint tests are passing.
